scipy/stats/_sensitivity_analysis.py

```diff
@@ -0,0 +1,364 @@
+    This is more efficient than using 2 random draw of Sobol'.
+    See sec. 5 from [1]_.
+
+    References
+    ----------
+    .. [1] Saltelli, A., P. Annoni, I. Azzini, F. Campolongo, M. Ratto, and
+       S. Tarantola. "Variance based sensitivity analysis of model
+       output. Design and estimator for the total sensitivity index."
+       Computer Physics Communications, 181(2):259-270,
+       :doi:`10.1016/j.cpc.2009.09.018`, 2010.
+    """
+    A_B = qmc.Sobol(d=2*d, seed=random_state, bits=64).random(n)
+
+    A, B = A_B[:, :d], A_B[:, d:]
+
+    if dists is not None:
+        for d_, dist in enumerate(dists):
+            dist_rng = NumericalInversePolynomial(
+                dist, random_state=random_state
+            )
+            A[:, d_] = dist_rng.ppf(A[:, d_])
+            B[:, d_] = dist_rng.ppf(B[:, d_])
+
+    return A, B
+
+
+def sample_AB(A: np.ndarray, B: np.ndarray) -> np.ndarray:
+    """AB matrix.
+
+    AB: columns of B into A. Shape (n*d, d). e.g in 2d
+    Take A and replace 1st column with 1st column of B. You have (n, d)
+    Then A and replace 2nd column with 2nd column of B. You have (n, d)
+    Concatenate to get AB. Which means AB is (2*n, d).
+    """
+    n, d = A.shape
+    AB = np.empty((int(d*n), d))
+    for i in range(d):
+        AB[i*n:(i+1)*n, :] = np.column_stack((A[:, :i], B[:, i], A[:, i+1:]))
+
+    return AB
+
+
+def saltelli_2010(
+    f_A: np.ndarray, f_B: np.ndarray, f_AB: np.ndarray
+) -> Tuple[np.ndarray,  np.ndarray]:
+    """Saltelli2010 formulation.
+
+    .. math::
+
+        S_i = \frac{1}{N} \sum_{j=1}^N
+        f(\mathbf{B})_j (f(\mathbf{AB}^{(i)})_j - f(\mathbf{A})_j)
+
+    .. math::
+
+        S_{T_i} = \frac{1}{N} \sum_{j=1}^N
+        (f(\mathbf{A})_j - f(\mathbf{AB}^{(i)})_j)^2
+
+    References
+    ----------
+    .. [1] Saltelli, A., P. Annoni, I. Azzini, F. Campolongo, M. Ratto, and
+       S. Tarantola. "Variance based sensitivity analysis of model
+       output. Design and estimator for the total sensitivity index."
+       Computer Physics Communications, 181(2):259-270,
+       :doi:`10.1016/j.cpc.2009.09.018`, 2010.
+    """
+    f_AB = f_AB.reshape(-1, f_A.shape[0])
+
+    var = np.var(np.vstack([f_A, f_B]))
+
+    s = np.mean(f_B * (f_AB - f_A.flatten()).T, axis=0) / var
+    st = 0.5 * np.mean((f_A.flatten() - f_AB).T ** 2, axis=0) / var
+
+    return s, st
+
+
+def sobol_indices(
+    *,
+    func: Callable,
+    n: IntNumber,
+    d: IntNumber,
+    dists: Optional[List[PINVDist]] = None,
+    l_bounds: Optional[npt.ArrayLike] = None,
+    u_bounds: Optional[npt.ArrayLike] = None,
+    random_state: SeedType = None
+):
+    """Global sensitivity indices of Sobol'.
+
+    Parameters
+    ----------
+    func : callable
+        Function to compute the Sobol' indices from. It's signature must be
+        ``func(x: ArrayLike) -> ArrayLike``, with ``x`` of shape ``(-1, d)``
+        and the output should have a shape ``(-1, 1)``.
```

Comment:
I did not yet mentioned any vectorisation. This would already be helpful as it is. But some people have temporal or spatial quantities of interest.

Classical example: you are looking at a river, the quantity of interest (QoI) is the water elevation Z. You have a model for Z which has 2 parameters, the flow rate Q and the friction coefficient (strickler) Ks. You could be interested to a single observation point along the river (e.g. there is a dam here), so you work with a scalar output: f(Q, Ks)=Z. But you could also be interested in having a look along the whole river (e.g. curvilinear abscise). Hence Z would be a vector.

We could either do the vectorisation in one go (cool if it works without killing readability). Or after we do the function evaluations we do a loop there.

Digression: it's not that critical in terms of speed. In real life, what's usually expensive is the function evaluation itself, the rest is usually marginal. And in real real life, people would use a surrogate model (usually Gaussian process or Polynomial chaos) as their model could take days to do a single evaluation.

Reply:
Oh! I just found it for the vect of output, I will fix my code. Not sure how I managed to mess this up the first time. Then only question will be how to use bootstrap here.

Reply:
I updated the code and added a test to show it works just fine 😃

There is also a concept of "aggregated" indices for the vector case, to come back to a single scalar. Can be done in further work I would say right now. (Although it's very simple, it's just a normalisation, something like `Siagg=sum(Vik)/sum(var(Vk)), with Vik = Sik*Vk`)